### PR TITLE
Go: support for aarch64

### PR DESCRIPTION
**Pull request by darmac**

### CODE REVIEW

**Comment by adamjstewart** [comment]
@hartzell did you have plans for aarch64 support?

**Comment by hartzell** [comment]
I'm kind of stuck right now, took a new job and it's getting in the way of me making contributions.

Starting simply, this is fine.

I think an easy but really useful improvement would be to control the choice of dependency provider by a variant, so that other platforms can use it easily.

I *think*, but haven't been able to dig in, that that in the long term there should be some go-bootstrapper abstraction that can be satisfied by various providers (go-binaries, gcc, externals, etc...).

File: var/spack/repos/builtin/packages/go/package.py [hunk 1]
```diff
@@ -35,16 +36,20 @@ class Go(Package):
...
     version('1.13.8', sha256='b13bf04633d4d8cf53226ebeaace8d4d2fd07ae6fa676d0844a688339debec34')
     version('1.13.7', sha256='e4ad42cc5f5c19521fbbbde3680995f2546110b5c6aa2b48c3754ff7af9b41f4')
     version('1.13.6', sha256='aae5be954bdc40bcf8006eb77e8d8a5dde412722bc8effcdaf9772620d06420c')
     version('1.13.5', sha256='27d356e2a0b30d9983b60a788cf225da5f914066b37a6b4f69d457ba55a626ff')
-    version('1.13.4',  sha256='95dbeab442ee2746b9acf0934c8e2fc26414a0565c008631b04addb8c02e7624')
+    version('1.13.4', sha256='95dbeab442ee2746b9acf0934c8e2fc26414a0565c008631b04addb8c02e7624')
     version('1.13.3', sha256='4f7123044375d5c404280737fbd2d0b17064b66182a65919ffe20ffe8620e3df')
     version('1.13.2', sha256='1ea68e01472e4276526902b8817abd65cf84ed921977266f0c11968d5e915f44')
     version('1.13.1', sha256='81f154e69544b9fa92b1475ff5f11e64270260d46e7e36c34aafc8bc96209358')
-    version('1.13', sha256='3fc0b8b6101d42efd7da1da3029c0a13f22079c0c37ef9730209d8ec665bf122')
```
Comment (hartzell) [inline var/spack/repos/builtin/packages/go/package.py:47]: I think that Spack has a policy of not trying to be a completist for releases, but only adding those that folks are actually using.  Once they get in, they're hard to get out (trust me, I know; I'm one of the folks who screams when someone removes that old thing that it turns out I depend on...).

Comment (darmac) [reply var/spack/repos/builtin/packages/go/package.py:47]: @hartzell 
Infact it's just refine the format to keep alignment:
```
    version('1.13', sha256='3fc0b8b6101d42efd7da1da3029c0a13f22079c0c37ef9730209d8ec665bf122')
```
=>
```
    version('1.13',    sha256='3fc0b8b6101d42efd7da1da3029c0a13f22079c0c37ef9730209d8ec665bf122')
```

the go@1.13 is not removed.

**Comment by darmac** [comment]
Is there any other suggestion to refine this patch?
@adamjstewart  @hartzell

**Comment by darmac** [comment]
Hi @adamjstewart  @hartzell 
We have some go package is blocked by this PR,
Is there any chance that we can use this patch to support some aarch64 go package?
Or we still need more time to think how to refine it?

**Review by adamjstewart — APPROVED** [review]
I'm fine with the changes in this PR if @hartzell is

**Review by hartzell — APPROVED** [review]
Looks good.  Sorry for mistaking the whitespace changes for new versions.

**Comment by sethrj** [comment]
@darmac Do you have any documentation on how this worked for you? My ~1-year-old machine (darwin-bigsur-m1) uses the wrong condition in `go` because `platform.machine` is `arm64` not `aarch64`; and working around that, `go` fails because `GOROOT_BOOTSTRAP` isn't set when gcc's go is a dependency.

**Comment by ecarre** [comment]
@sethrj I run spack (current develop branch) on a Taishan server (Kunpeng 920 CPU) give me :
```
==> go-bootstrap: Executing phase: 'install'
==> [2022-04-27-10:46:11.052140] '/usr/bin/bash' 'make.bash'
# Building C bootstrap tool.
cmd/dist
go tool dist: unknown architecture: aarch64
```

The fix from @darmac was working well for me, when I try to put it back I succeed to build go 1.18
```
==> Installing go-1.18-qphmk365khvn2kxxj45nqxc5h3y3ab2g
==> No binary for go-1.18-qphmk365khvn2kxxj45nqxc5h3y3ab2g found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/38/38f423db4cc834883f2b52344282fa7a39fbb93650dc62a11fdf0be6409bdad6.tar.gz
==> No patches needed for go
==> go: Executing phase: 'install'
==> go: Successfully installed go-1.18-qphmk365khvn2kxxj45nqxc5h3y3ab2g
  Fetch: 1.08s.  Build: 6m 43.92s.  Total: 6m 44.99s.
```

**Pull request merged**